### PR TITLE
[shelly] BLU Motion,  optimize ShellyManager for BLU devices

### PR DESCRIPTION
**Pull request by markus7017**
The PR includes the following enhancements and fixes. Most changes are related to device discovery and initialization as well as BLU motion support. The other topics are small fixes, we were able to catch a bunch of them.
- #15359: Shelly BLU Motion is supported
- #15396: Cleanup in Shelly Manager for actions and firmware upgrade, which currently can’t be supported by the binding.
- README updated

Signed-off-by: Markus Michels <markus7017@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties [hunk 1]
```diff
@@ -42,7 +42,7 @@ message.event.triggered = Event triggered: {0}
 message.event.filtered = Event filtered: {0}
 message.coap.init.failed = Unable to start CoIoT: {0}
 message.discovery.disabled = Device is marked as non-discoverable, will be skipped
-message.discovery.protected = Device {0} reported 'Access defined' (missing userid/password or incorrect).
+message.discovery.protected = Device {0} is protected and reports 'Access defined', check userId/password.
```
Comment (lsiepel) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:45]: Typo? defined vs denied ?

File: bundles/org.openhab.binding.shelly/README.md [hunk 2]
```diff
@@ -1481,6 +1494,18 @@ See notes on discovery of Shelly BLU devices above.
 |         | lux         | Number   | yes       | Brightness in Lux reported by the integrated sensor                               |
 |         | lastUpdate  | DateTime | yes       | Timestamp of the last update (any sensor value changed)                           |
 
+### Shelly BLU Motion Sensor (thing-type: shellyblumotion)
```
Comment (lsiepel) [inline bundles/org.openhab.binding.shelly/README.md:1497]: The same table was included before in line 1470.

**Comment by markus7017** [comment]
@lsiepel changes applied

**Comment by markus7017** [comment]
@lsiepel anything missing?

**Comment by lsiepel** [comment]
> @lsiepel anything missing?

I don't have shelly devices and this binding is not known area to me, so i guess it's better for others to review.

**Comment by markus7017** [comment]
@openhab/add-ons-maintainers Could somebody please pick-up this PR

**Comment by lolodomo** [comment]
@markus7017 : why you don't create different small PRs for each fixed issue or new enhancement ? I am sure you will get quicker feedback. Recommendation for new PRs.

**Comment by markus7017** [comment]
@lolodomo 90% (about 270 lines) are dependent on BLU motion, required changes to general initialization and decided to do
some cleanup related to that. The remaining 30 lines are related to small fixes I found along the way.

Creating a separate PR for every logic change or bug fix creates a lot of overhead and makes it hard to manage PRs in parallel. Due to longer response times I  have already a bunch of new changes I kept out of this PR to keep it small. Allterco is delivering a lot of nee devices, which change a) in direct changes b) require changes to the generic logic and c) maybe bugs need to be fixed. So it's unrealistic to serialize those changes, complete one by the other and the create a dozen 50 line PRs. 

Do I need to remove the 30 lines bug fixes and crate another PR for those?

**Comment by lolodomo** [comment]
> Do I need to remove the 30 lines bug fixes and crate another PR for those?

Let's keep them.
My comment was only an advice for the future.

**Comment by lolodomo** [comment]
Note that there is one file in conflict.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 3]
```diff
@@ -1014,12 +1006,19 @@ protected void initializeThingConfig() {
         config.serviceName = getString(properties.get(PROPERTY_SERVICE_NAME));
         config.localIp = bindingConfig.localIP;
         config.localPort = String.valueOf(bindingConfig.httpPort);
-        if (config.userId.isEmpty() && !bindingConfig.defaultUserId.isEmpty()) {
+        if (!profile.isGen2 && config.userId.isEmpty() && !bindingConfig.defaultUserId.isEmpty()) {
+            // // Gen2 has hard coded user "admin"
```
Comment (lolodomo) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:1010]: Minor: please remove one of the //

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 4]
```diff
@@ -1014,12 +1006,19 @@ protected void initializeThingConfig() {
...
+        if (config.password.isEmpty() && !bindingConfig.defaultPassword.isEmpty()) {
             config.password = bindingConfig.defaultPassword;
-            logger.debug("{}: Using userId {} from bindingConfig", thingName, config.userId);
+            logger.debug("{}: Using default password from bindingConfig (userId={})", thingName, config.userId);
         }
-        if (config.updateInterval == 0) {
+
+        if (config.updateInterval == 0)
+
+        {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:1021]: Please revert

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Comment by lolodomo** [comment]
@markus7017 : I reviewed fully your PR. There is one important comment to consider. Please also note that a file conflict has to be resolved.

**Comment by lolodomo** [comment]
@markus7017 : I do not know if you are available now. There is a new milestone build that will start in few hours. If you can consider my comments and resolve the conflict, I will try to merge your PR just in time for the new milestone.

**Comment by markus7017** [comment]
@lolodomo I was busy the last weeks, just working on this PR and 2 important fixes

File: bundles/org.openhab.binding.shelly/README.md [hunk 5]
```diff
@@ -1465,6 +1467,17 @@ See notes on discovery of Shelly BLU devices above.
 |         | lowBattery    | Switch   | yes       | Low battery alert (< 20%)                                                           |
 | device  | gatewayDevice | String   | yes       | Shelly forwarded last status update (BLU gateway), could vary from packet to packet |
 
+## Shelly BLU Motion Sensor (thing-type: shellyblumotion)
+
+See notes on discovery of Shelly BLU devices above.
+
+| Group   | Channel       | Type     | read-only | Description                                                                         |
+| ------- | ------------- | -------- | --------- | ----------------------------------------------------------------------------------- |
+| sensors | motion        | Switch   | yes       | ON: Motion detected                            d                                    |
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:1476]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 4 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:1021]: fixed

**Comment by markus7017** [comment]
@lolodomo one question: pom.xml still includes a dependency on org.eclipse.jetty.websocket. Is this still required? Should I keep it for backward compatibility?

**Comment by jlaur** [comment]
> > Do I need to remove the 30 lines bug fixes and crate another PR for those?
> 
> Let's keep them. My comment was only an advice for the future.

[...]

> I was busy the last weeks, just working on this PR and 2 important fixes

I will not interfere with the agreement between you and @lolodomo, but just add that if those fixes are important for 4.0 as well, I would advise you to create a separate PR with them so we can cherry-pick them into `4.0.x`.

Also, I noticed many issues mentioned in the PR description. @markus7017, can you link those issues that will be closed by this PR? When they are linked, they will be automatically closed when the PR is merged, so it would make it easier to manage.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java [hunk 6]
```diff
@@ -389,14 +395,23 @@ public static String extractFwVersion(@Nullable String version) {
...
+                    : GEN2_VERSION_PATTERN.matcher(vers);
             if (matcher.find()) {
                 return matcher.group(0);
             }
         }
         return "";
     }
 
+    public static boolean isGeneration2(String thingType) {
+        return thingType.startsWith("shellyplus") || thingType.startsWith("shellypro") | isBluSeries(thingType);
```
Comment (maniac103) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:408]: While a bitwise or also will work, I think it's still worth to fix the typo: `|` -> `||`

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 7]
```diff
@@ -570,9 +563,8 @@ protected void refreshStatus() {
             } else if (isWatchdogStarted()) {
                 if (!isWatchdogExpired()) {
                     logger.debug("{}: Ignore API Timeout on {} {}, retry later", thingName, res.method, res.url);
-                } else {
-                    if (isThingOnline()) {
-                        status = "offline.status-error-watchdog";
+                    if (profile.alwaysOn) { // suspress for battery powered sensors
```
Comment (maniac103) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:566]: `suppress`

**Comment by lolodomo** [comment]
> @lolodomo one question: pom.xml still includes a dependency on org.eclipse.jetty.websocket. Is this still required? Should I keep it for backward compatibility?

That's strange, why did you do that first ?? You should rely on current version of Jetty "packaged" with OH server. We are currently at 9.4.52.v20230823 after upgrading to Karaf 4.4.4.
https://github.com/openhab/openhab-core/pull/3814

**Comment by lolodomo** [comment]
@markus7017 : you marked all my comments as resolved without pushing any changes !

**Comment by markus7017** [comment]
> Also, I noticed many issues mentioned in the PR description. @markus7017, can you link those issues that will be closed by this PR? When they are linked, they will be automatically closed when the PR is merged, so it would make it easier to manage.

Hmm, I reviewed the above list, which is correct. Most of them were tiny fixes, one line or so.
#15393 is not a must, more a simplified handling for the user
#15398 caused duplicate entries in the Inbox
#15397 is a must for firmware 1.0+ and password protected devices due to strict RFC-based implementation
#15932 is cosmetics
#15350 an enhancement for new device
#15396 improved user experience
#15395 caused initialization failure for the Plus HT sensor
#15394 enhancement
#15393 fixes an initialization problem for TRV

and this one includes 2 fixes for resource leaks when discovery handler fails (NOE+IPv6)

All of them would be useful to fixes (serious) issues or little improvements. So would make sense also to get into 4.0. 
@jlaur @lolodomo Let me now to to proceed. I could create another PR and fiddle out those lines included to the agreed list.

**Comment by markus7017** [comment]
I know, I need to do a rebase and to double check that all resource leak fixes are included

**Comment by jlaur** [comment]
> All of them would be useful to fixes (serious) issues or little improvements. So would make sense also to get into 4.0. @jlaur @lolodomo Let me now to to proceed. I could create another PR and fiddle out those lines included to the agreed list.

It's a matter of reducing risk as it would be counter-productive to introduce new regressions in a hotfix. 😉 Therefore, at first glance I would skip all those where you mentioned "not a must", "enhancement", "improved" and "cosmetics". They can be kept in this PR as already agreed with @lolodomo and won't be cherry-picked for 4.0.

You could group the fixes that are conflicting with each other in the same PR, but other than that I would actually propose one PR per fix (with corresponding issue linked). I don't mind cherry-picking multiple commits, and release notes would be nicer because each fix could be better described in the PR title. The review process should also be faster, because doubts/discussions for one PR won't block the other PR's.

This is just my proposition, and if you and @lolodomo would prefer another approach, that's fine. I don't think a PR like this one in its entirety has a very good chance to be cherry-picked though.

**Comment by markus7017** [comment]
I added issues for the 2 fixes we discovered
- #15795 Fix resource leak on exception while processing mDNS result
- #15796 NPE on IPv6 discovrxy result
-

**Comment by markus7017** [comment]
How do we handle 4.0?

**Comment by jlaur** [comment]
> How do we handle 4.0?

Two possibilities:
1. When small PR's are merged into `main`, their commits can be cherry-picked into `4.0.x` (unless there are other changes merged already that could cause conflicts).
2. The other possibility is dedicated PR's targeting `4.0.x`. This option should only be used if it's not possible to cherry-pick.

**Comment by markus7017** [comment]
ok, I'll create with the fixes
which gets merged -> removes code here
then we could merge this one to main
and cherry pick the PR for 4.0

**Comment by markus7017** [comment]
created #15798, which should be merged to main and 4.0

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java [hunk 8]
```diff
@@ -218,7 +223,8 @@ public void initFromThingType(String name) {
         isHT = thingType.equals(THING_TYPE_SHELLYHT_STR) || thingType.equals(THING_TYPE_SHELLYPLUSHT_STR);
         isDW = thingType.equals(THING_TYPE_SHELLYDOORWIN_STR) || thingType.equals(THING_TYPE_SHELLYDOORWIN2_STR)
                 || thingType.equals(THING_TYPE_SHELLYBLUDW_STR);
-        isMotion = thingType.startsWith(THING_TYPE_SHELLYMOTION_STR);
+        isMotion = thingType.startsWith(THING_TYPE_SHELLYMOTION_STR)
+                || thingType.equals(THING_TYPE_SHELLYBLUMOTION_STR);
```
Comment (ickeundso) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:227]: the thingType.equals(THING_TYPE_SHELLYBLUMOTION_STR); is redundant, every time when equals evaluates to true the startsWith will evaluate to true too. So if startsWith is sufficient the || equals can be omitted.

File: bundles/org.openhab.binding.shelly/README.md [hunk 9]
```diff
@@ -120,7 +122,7 @@ The binding provides the same feature set across all devices as good as possible
 | ----------------- | ------------------------------------------------------ | --------- |
 | shellyblubutton   | Shelly BLU Button 1                                    | SBBT      |
 | shellybludw       | Shelly BLU Door/Windows                                | SBDW      |
-
+| shellyblumotion   | Shelly BLU Motion                                      | SBM       |
```
Comment (ickeundso) [inline bundles/org.openhab.binding.shelly/README.md:125]:
```suggestion
| shellyblumotion   | Shelly BLU Motion                                      | SBMO      |
```

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java (hunk 8 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:227]: Note, it's THING_TYPE_SHELLYMOTION_STR and THING_TYPE_SHELLY**BLU**MOTION_STR

File: bundles/org.openhab.binding.shelly/README.md (hunk 9 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:125]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java [hunk 10]
```diff
@@ -118,12 +115,18 @@ public class ShellyDeviceProfile {
...
-
         initialized = false;
+        if (device != null) {
+            this.device = device;
+        }
 
         initFromThingType(thingType);
+        if (device != null) {
+            this.device = device;
+        }
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java:129]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiClient.java [hunk 11]
```diff
@@ -148,6 +148,11 @@ public Shelly2ApiClient(String thingName, ShellyThingConfiguration config, HttpC
         MAP_ROLLER_STATE.put(SHELLY2_RSTATE_STOPPED, SHELLY_RSTATE_STOP);
         MAP_ROLLER_STATE.put(SHELLY2_RSTATE_CALIB, SHELLY2_RSTATE_CALIB); // Gen2-only
     }
+    protected static final Map<String, String> MAP_PROFILE = new HashMap<>();
+    static {
+        MAP_ROLLER_STATE.put(SHELLY_CLASS_RELAY, SHELLY2_PROFILE_RELAY);
+        MAP_ROLLER_STATE.put(SHELLY_CLASS_ROLLER, SHELLY2_PROFILE_COVER);
+    }
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiClient.java:155]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 3 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java (hunk 6 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyDeviceProfile.java]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 7 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java]: fixed

**Comment by markus7017** [comment]
@lolodomo This is the left-over based one the 2 PRs before :-)

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties [hunk 12]
```diff
@@ -122,6 +122,7 @@ thing-type.shelly.shellypro4pm.description = Shelly Pro 4PM - 4xRelay Switch wit
 # BLU devices
 thing-type.shelly.shellyblubutton.description = Shelly BLU Button
 thing-type.shelly.shellybludw.description = Shelly BLU Door/Window Sensor
+ting-type.shelly.shellyblumotion.description = Shelly BLU Motion Sensor
```
Comment (lolodomo) [inline bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:125]: a "h" is missing: thing and not ting

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/manager/ShellyManagerActionPage.java [hunk 13]
```diff
@@ -377,6 +377,9 @@ public static Map<String, String> getActions(ShellyDeviceProfile profile) {
 
         list.put(ACTION_RES_STATS, "Reset Statistics");
         list.put(ACTION_RESTART, "Reboot Device");
```
Comment (lolodomo) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/manager/ShellyManagerActionPage.java:379]: I guess this line has to be removed.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java [hunk 14]
```diff
@@ -1019,7 +1019,10 @@ protected void initializeThingConfig() {
             config.password = bindingConfig.defaultPassword;
             logger.debug("{}: Using default password from bindingConfig (userId={})", thingName, config.userId);
         }
-        if (config.updateInterval == 0) {
+
+        if (config.updateInterval == 0)
+
+        {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:1025]: Please undo that change

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties (hunk 12 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/i18n/shelly.properties:125]: fixed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/manager/ShellyManagerActionPage.java (hunk 13 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/manager/ShellyManagerActionPage.java:379]: yes

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java (hunk 14 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java:1025]: done

**Comment by markus7017** [comment]
@lolodomo requested changes applied

File: bundles/org.openhab.binding.shelly/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.shelly/README.md:1476]: No, still not fixed

**Comment by lolodomo** [comment]
> @lolodomo requested changes applied

One comment in README you marked as resolved but you did not fix.

Please add these kind of entries in the future for an automatic closing of issues.

Closes #15359 
Closes #15396

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by markus7017** [comment]
> > @lolodomo requested changes applied
> 
> One comment in README you marked as resolved but you did not fix.
> 
> Please add these kind of entries in the future for an automatic closing of issues.
> 
> Closes #15359 Closes #15396

@lolodomo 
Isn't that done automatically when I add the issues under "Development" on the right side?
In addition Closes #15909 

changes applied

**Comment by markus7017** [comment]
@lolodomo thank you!!

**Comment by lolodomo** [comment]
> @lolodomo thank you!!

As you can notice, small PR = faster review

**Pull request merged**